storage_service/locations/models/dataverse.py

```diff
@@ -112,63 +212,113 @@ def move_to_storage_service(self, src_path, dest_path, dest_space):
+                # Tabular Data File. The .tab file format is downloaded as a
+                # 'bundle' from Dataverse. This bundle provides multiple
+                # representations in different formats of the same data.
+                #
+                # A bundle has the property of being a zip file when pulled
+                # down by the storage service. We want to extract this bundle
+                # below, and allow Archivematica to process other zip files as
+                # it would normally (as configured) in the transfer workflow.
+                #
+                # Integrity checks are completed by the Dataverse
+                # microservices.
+                zipped_bundle = True
+                download_path = os.path.join(
+                    dest_path, file_entry["label"][:-4] + ".zip"
+                )
+                url = "https://{}/api/access/datafile/bundle/{}".format(
+                    self.host, entry_id
+                )
+            LOGGER.debug("URL: %s, params: %s", url, params)
+            response = requests.get(url, params=params, stream=True)
+            with open(download_path, "wb") as f:
+                for chunk in response.iter_content(8192):
+                    f.write(chunk)
+            if zipped_bundle and ".zip" in download_path:
+                # The bundle .zip itself is ephemeral, and so once downloaded
+                # unzip and remove the container here.
+                LOGGER.info("Bundle downloaded. Deleting.")
+                self.extract_and_remove_bundle(dest_path, download_path)
 
         # Add Agent info
         agent_info = [
             {
-                'agentIdentifierType': 'URI',
-                'agentIdentifierValue': self.agent_identifier,
-                'agentName': self.agent_name,
-                'agentType': self.agent_type,
+                "agentIdentifierType": "URI",
+                "agentIdentifierValue": self.agent_identifier,
+                "agentName": self.agent_name,
+                "agentType": self.agent_type,
             }
         ]
-        agentjson_path = os.path.join(dest_path, 'metadata', 'agents.json')
-        with open(agentjson_path, 'w') as f:
-            json.dump(agent_info, f)
+        agentjson_path = os.path.join(dest_path, "metadata", "agents.json")
+        with open(agentjson_path, "w") as f:
+            json.dump(agent_info, f, sort_keys=True, indent=4,
+                      separators=(',', ': '))
+
+    @staticmethod
+    def extract_and_remove_bundle(dest_path, bundle_path):
+        try:
+            unzipper = zipfile.ZipFile(bundle_path, 'r')
+        except zipfile.BadZipfile as e:
+            # Log the error and return without extracting the bundle.
+            # Archivematica may still be able to work with the data returned.
+            LOGGER.info("Bundle '%s' error: %s", bundle_path, e)
+            return
+        unzipper.extractall(dest_path)
+        unzipper.close()
```

Comment:
Can you use a context manager in the above call to `ZipFile` so that you don't have to manually close `unzipper`? I.e., `with zipfile.ZipFile(bundle_path, 'r') as unzipper`.

Reply:
:+1: